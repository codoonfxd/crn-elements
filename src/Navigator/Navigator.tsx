/*
 * @Author: JaneEyre(lsy@codoon.com)
 * @Date: 2019-04-23 11:19:22
 * @Last Modified by: JaneEyre(lsy@codoon.com)
 * @Last Modified time: 2019-06-12 13:47:28
 * @Content: Navigator
 */
import * as React from 'react'
import {
  View,
  ViewStyle,
  TouchableOpacity,
  TextStyle,
  Animated,
  Image,
} from 'react-native'

import styles from './style'
import { isClassComponent, isFunctionComponent } from '../lib/utils'
import { IS_IPHONE_X } from '../lib/device'

export type IButtonType = React.ReactNode | React.Component
export interface INavigatorProps {
  // navigator title
  title: string
  // custom navigator style
  style?: ViewStyle
  // custom navigator title style
  titleStyle?: TextStyle
  // wether absolute positioning is supported
  absolute?: boolean
  // navigator left button config
  rightButtonStyle?: ViewStyle
  // right button
  rightButton?: IButtonType
  // navigator left button style
  leftButtonStyle?: ViewStyle
  // left button
  leftButton?: IButtonType
  // left button press callback
  onPress?: () => void
}

const ARROW_BACK =
  // tslint:disable-next-line: max-line-length
  'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEgAAABICAQAAAD/5HvMAAAB50lEQVRo3u3ZzSvEQRgHcGu9tCQuckIOchDlhrgpXOTC88x63U2L4iCFcNhyUQ4ObhK1LtqQkpLigBwk7RG5kCRFDuTdsktJnjm4+M5hvvsPfHpmfvM8MxsTY2NjY2NjY2Pzx4Qd3E+73GEIJ+jkGQpHfuw3gONP4IVPjhGkpmRe++bASa1pvPOTEyWh9pLKoNBvzgdoE8JpzuIjkfOq6gGcxjw+FTlPbgZwuIguRc6tuxKxd8roRuLQVUMxglNFdyLnTOUDOFTHT+JiHbqzEdVp41eRs+dJR3B6xaUK04Y3BfFljcgcXuxORIwXExrOVND5//08jgMazihgqboTeUnkvKlezHixLnKeVQukhfKsuFT3XIPhlGv2jhc08fCg3EIBX9ZXoxgQK3QHA6kSzdncCZuaeUoEPUKmwkh6XLSiGVTbQSRfPM9pFm4IdRWM5UkNaTzsQO2lMc2JFPDHoUjDGtJyjwt1KnXRm3wx9KWCSO5mehHrFFIZqKOylh7EKh035aCqVMG3Ium8oQDWUPhaJF1zKYpUyBdy26Vq1MLl0on82KAU6lzKpAN51mYCkTzpvC9WaQs2nPhSadugJ70oKYlWDXr0jM4CCTRvECeSoJOmDeJ83v1Vn0F/LdjY2NjY2NgYlneNdJiYTAB1AAAAAABJRU5ErkJggg=='

class Navigator extends React.Component<INavigatorProps> {
  static defaultProps: INavigatorProps = {
    title: '',
    style: {},
    titleStyle: {},
    absolute: false,
  }

  /**
   * get the button child component according to the child type
   * @param { object } button
   * @returns { object }
   */
  getButtonComponent = (button: IButtonType): IButtonType => {
    if (React.isValidElement(button)) {
      return button
    } else if (isClassComponent(button) || isFunctionComponent(button)) {
      const Child = button as React.ComponentClass
      return <Child />
    } else {
      console.error(
        'button children is not a valid react node or react component..'
      )
      return null
    }
  }

  render() {
    const {
      title,
      style,
      titleStyle,
      absolute,
      rightButton,
      leftButton,
      leftButtonStyle,
      rightButtonStyle,
      onPress,
    } = this.props

    const defaultLeftButton = (
      <TouchableOpacity
        style={[styles.leftButtonContainer, leftButtonStyle]}
        onPress={onPress}
      >
        <Image source={{ uri: ARROW_BACK }} style={styles.arrowBack} />
      </TouchableOpacity>
    )

    return (
      <Animated.View
        style={[styles.wrapper, absolute ? styles.absolute : null, style]}
      >
        {absolute && IS_IPHONE_X && (
          <View
            style={{
              height: 30,
              backgroundColor: 'transparent',
            }}
          />
        )}
        <Animated.View style={[styles.innerWrapper, style]}>
          {leftButton ? (
            <View style={[styles.leftButtonContainer, leftButtonStyle]}>
              {this.getButtonComponent(leftButton)}
            </View>
          ) : (
            defaultLeftButton
          )}

          <View style={styles.titleContainer}>
            <Animated.Text style={[styles.titleStyle, titleStyle]}>
              {title}
            </Animated.Text>
          </View>

          <View style={[styles.rightButtonContainer, rightButtonStyle]}>
            {rightButton && this.getButtonComponent(rightButton)}
          </View>
        </Animated.View>
      </Animated.View>
    )
  }
}

export default Navigator
