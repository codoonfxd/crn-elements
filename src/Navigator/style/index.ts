import { StyleSheet, Dimensions } from 'react-native'
import { DEVICE_WIDTH } from '../../lib/constant'

const style = StyleSheet.create({
  wrapper: {
    height: 50,
    backgroundColor: '#fff',
    flexDirection: 'column',
  },
  innerWrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    height: 50,
    borderBottomWidth: 1,
    backgroundColor: '#fff',
    borderColor: '#f2f2f2',
    borderStyle: 'solid',
  },
  arrowBack: {
    width: 20,
    height: 20,
  },
  titleContainer: {
    flex: 2,
    justifyContent: 'center',
    alignItems: 'center',
  },
  titleStyle: {
    fontSize: 14,
    color: '#222',
    fontWeight: 'bold',
  },
  leftButtonContainer: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    paddingLeft: 16,
  },
  rightButtonContainer: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'flex-end',
    paddingRight: 16,
  },
  absolute: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: DEVICE_WIDTH,
    zIndex: 100,
  },
})

export default style
