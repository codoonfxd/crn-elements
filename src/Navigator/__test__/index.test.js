import React from "react";
import { View, Text } from "react-native";

import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import Navigator from "../Navigator";

describe("<Navigator />", () => {
  it("snapshot", () => {
    const tree = renderer.create(<Navigator title="title" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render title when the title prop is passed", () => {
    const title = "title";
    const wrapper = shallow(<Navigator title={title} />);
    expect(wrapper.contains(title)).toBe(true);
  });

  it("should render left button correctly when the leftbutton as a react element is passed", () => {
    const leftButton = (
      <View>
        <Text>leftButton</Text>
      </View>
    )
    const wrapper = shallow(<Navigator leftButton={leftButton} />);
    expect(wrapper.contains(leftButtonChild)).toBe(true);
  });

  it("should render left button correctly when the leftbutton as a react class component is passed", () => {
    class leftButtonComp extends React.Component {
      render() {
        return <Text>xxx</Text>;
      }
    }
    const wrapper = shallow(<Navigator leftButton={leftButtonComp} />);
    expect(wrapper.find(leftButtonComp)).toHaveLength(1);
  });

  it("should render left button correctly when the leftbutton as a react function component is passed", () => {
    const leftButtonComp = () => (
      <Text>xxx</Text>
    )
    const wrapper = shallow(<Navigator leftButton={leftButtonComp} />);
    expect(wrapper.find(leftButtonComp)).toHaveLength(1);
  });

  it("should render right button correctly when the rightbutton is passed", () => {
    const rightButtonChild = (
      <View>
        <Text>rightButton</Text>
      </View>
    );
    const wrapper = shallow(<Navigator rightButton={rightButtonChild} />);
    expect(wrapper.contains(rightButtonChild)).toBe(true);
  });
});
