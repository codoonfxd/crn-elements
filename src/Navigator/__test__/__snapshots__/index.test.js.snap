// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigator /> snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "borderBottomWidth": 1,
      "borderColor": "#f2f2f2",
      "borderStyle": "solid",
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
        "paddingLeft": 16,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 2,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#222",
          "fontSize": 14,
          "fontWeight": "bold",
        }
      }
    >
      title
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
        "paddingRight": 16,
      }
    }
  />
</View>
`;
